jsonrpc/ipc_all.go

```diff
@@ -0,0 +1,61 @@
+package jsonrpc
+
+import (
+	"errors"
+	"net"
+	"os"
+	"path/filepath"
+)
+
+const (
+	// On Linux, path is 108 bytes in size.
+	// http://man7.org/linux/man-pages/man7/unix.7.html
+	maxUnixPathBytes = 108
+	// fileModePerms represents the permission mode for directories, which is set to 751.
+	// This mode is stricter than the common default of 755 for directories.
+	fileModePerms = 0o751
+	// socketPerms represents the permission mode for socket files, set to 600. It allows
+	// read and write access for the owner only, ensuring that socket is accessible only to the intended user.
+	socketPerms = 0o600
+)
+
+var errPathTooLong = errors.New("path too long")
+
+// createListener creates a Unix domain socket at the given path and sets proper permissions.
+func createListener(endpoint string) (net.Listener, error) {
+	var (
+		l   net.Listener
+		err error
+	)
+	// path + terminator
+	if len(endpoint)+1 > maxUnixPathBytes {
+		return nil, errPathTooLong
+	}
+	err = preparePath(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	l, err = net.Listen("unix", endpoint)
+	if err != nil {
+		return nil, err
+	}
+	// Set permissions for the socket file to read and write for the owner only (0o600)
+	err = os.Chmod(endpoint, socketPerms)
+	return l, err
+}
+
+// preparePath ensures the existence of directories leading to provided path
+// and removes existing file if any.
+func preparePath(path string) error {
+	var err error
+	err = os.MkdirAll(filepath.Dir(path), fileModePerms)
+	if err != nil {
+		return err
+	}
+	// Remove any existing file at the specified path.
+	err = os.Remove(path)
```

Comment:
[Here](https://github.com/tailscale/tailscale/blob/e78cb9aeb3b32ee97991710c72eb979ebad61477/safesocket/unixsocket.go#L27-L45) is an alternative approach that seems to achieve the best of both worlds: if the socket exists and is in use, shut down with an error, otherwise use the existing socket. Thoughts?

Reply:
Sounds good to me (wasn't aware that dial is able to detect stale socket, thought it will simply connect) expect using the existing socket. It has to be recreated in order to net.Listen.